modules/javafx.controls/src/main/java/javafx/scene/control/skin/MenuBarSkin.java

```diff
@@ -389,62 +276,141 @@ public MenuBarSkin(final MenuBar control) {
+                                            moveToMenu(Direction.PREVIOUS, false); // just move the selection bar
+                                        } else {
+                                            moveToMenu(Direction.NEXT, false); // just move the selection bar
+                                        }
+                                        ev.consume();
+                                        return;
+                                    }
+                                    if (isRTL) {
+                                        moveToMenu(Direction.PREVIOUS, true);
+                                    } else {
+                                        moveToMenu(Direction.NEXT, true);
+                                    }
+                                }
+                                ev.consume();
+                                break;
+                            }
+                            case DOWN:
+                            //case SPACE:
+                            //case ENTER:
+                                // RT-18859: Doing nothing for space and enter
+                                if (control.getScene().getWindow().isFocused()) {
+                                    if (focusedMenuIndex != -1) {
+                                        Menu menuToOpen = getSkinnable().getMenus().get(focusedMenuIndex);
+                                        showMenu(menuToOpen, true);
+                                        ev.consume();
+                                    }
+                                }
+                                break;
+                            case ESCAPE:
+                                unSelectMenus();
+                                ev.consume();
+                                break;
+                        default:
+                            break;
+                        }
+                    }
+                });
+
+                // When we click else where in the scene - menu selection should be cleared.
+                sceneListenerHelper.addEventFilter(scene, MouseEvent.MOUSE_CLICKED, (ev) -> {
+                    Bounds containerScreenBounds = container.localToScreen(container.getLayoutBounds());
+                    if ((containerScreenBounds == null) || !containerScreenBounds.contains(ev.getScreenX(), ev.getScreenY())) {
+                        unSelectMenus();
+                    }
+                });
+
+                // When the parent window looses focus - menu selection should be cleared
+                sceneListenerHelper.addChangeListener(scene.windowProperty(), true, (sr, oldw, w) -> {
+                    if (windowFocusHelper != null) {
+                        windowFocusHelper.disconnect();
+                        windowFocusHelper = null;
+                    }
+
+                    if (w != null) {
+                        windowFocusHelper = sceneListenerHelper.addChangeListener(w.focusedProperty(), true, (s, p, focused) -> {
```

Comment:
```suggestion
                        windowFocusHelper = sceneListenerHelper.addChangeListener(w.focusedProperty(), true, focused -> {
```

Reply:
my version does not create extra object(s).